test/e2e/cluster_upgrade_runtimesdk.go

```diff
@@ -0,0 +1,226 @@
+
+		if input.ControlPlaneMachineCount == nil {
+			controlPlaneMachineCount = 1
+		} else {
+			controlPlaneMachineCount = *input.ControlPlaneMachineCount
+		}
+
+		if input.WorkerMachineCount == nil {
+			workerMachineCount = 2
+		} else {
+			workerMachineCount = *input.WorkerMachineCount
+		}
+
+		// Setup a Namespace where to host objects for this spec and create a watcher for the Namespace events.
+		namespace, cancelWatches = setupSpecNamespace(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder)
+		clusterResources = new(clusterctl.ApplyClusterTemplateAndWaitResult)
+	})
+
+	It("Should create and upgrade a workload cluster", func() {
+		By("Deploy Test Extension")
+		testExtensionDeploymentTemplate, err := os.ReadFile(testExtensionPath) //nolint:gosec
+		Expect(err).ToNot(HaveOccurred(), "Failed to read the extension config deployment manifest file")
+		testExtensionDeployment := strings.ReplaceAll(string(testExtensionDeploymentTemplate), "${SERVICE_NAMESPACE}", namespace.Name)
+		Expect(testExtensionDeployment).ToNot(BeEmpty(), "Test Extension deployment manifest file should not be empty")
+		Expect(input.BootstrapClusterProxy.Apply(ctx, []byte(testExtensionDeployment), "--namespace", namespace.Name)).To(Succeed())
+
+		By("Deploy Test Extension ExtensionConfig")
+		ext = extensionConfig(specName, namespace)
+		err = input.BootstrapClusterProxy.GetClient().Create(ctx, ext)
+		Expect(err).ToNot(HaveOccurred(), "Failed to create the extension config")
+
+		By("Creating a workload cluster")
+
+		clusterctl.ApplyClusterTemplateAndWait(ctx, clusterctl.ApplyClusterTemplateAndWaitInput{
+			ClusterProxy: input.BootstrapClusterProxy,
+			ConfigCluster: clusterctl.ConfigClusterInput{
+				LogFolder:                filepath.Join(input.ArtifactFolder, "clusters", input.BootstrapClusterProxy.GetName()),
+				ClusterctlConfigPath:     input.ClusterctlConfigPath,
+				KubeconfigPath:           input.BootstrapClusterProxy.GetKubeconfigPath(),
+				InfrastructureProvider:   clusterctl.DefaultInfrastructureProvider,
+				Flavor:                   pointer.StringDeref(input.Flavor, "upgrades"),
+				Namespace:                namespace.Name,
+				ClusterName:              fmt.Sprintf("%s-%s", specName, util.RandomString(6)),
+				KubernetesVersion:        input.E2EConfig.GetVariable(KubernetesVersionUpgradeFrom),
+				ControlPlaneMachineCount: pointer.Int64Ptr(controlPlaneMachineCount),
+				WorkerMachineCount:       pointer.Int64Ptr(workerMachineCount),
+			},
+			WaitForClusterIntervals:      input.E2EConfig.GetIntervals(specName, "wait-cluster"),
+			WaitForControlPlaneIntervals: input.E2EConfig.GetIntervals(specName, "wait-control-plane"),
+			WaitForMachineDeployments:    input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
+			WaitForMachinePools:          input.E2EConfig.GetIntervals(specName, "wait-machine-pool-nodes"),
+		}, clusterResources)
+
+		By("Upgrading the Cluster topology")
```

Comment:
nit: add a comment specifying we are running an upgrade to test all the lifecycle hooks

Reply:
Done, although this code will change a lot with Yuvaraj's PR to go step-by-step through the cluster lifecycle